### PR TITLE
Added oversampling and undersampling

### PR DESCRIPTION
**Pull request by michail-tzoufras**
Changes:

1) Added functions for oversampling and undersampling the imbalanced dataset
2) Moved data processing functions to a separate file.
3) Edited the way I call the deep learning model with embeddings.

Best regards,
Michail

### CODE REVIEW

File: Insight_Project_Framework/embeddings_DL.py [hunk 1]
```diff
@@ -45,15 +45,15 @@ def __init__(self, vocabulary_sizes, max_length, num_ordinal_features):
         self.merged_model = Model(inputs=embedding_inputs+ordinal_inputs, outputs=output)
 
 
-    def __call__(self, _input_data, _labels, quiet = False):
+    def train(self, _input_data, _labels, quiet = False):
         """compiles the model, fits the _input_data to the _labels, and evaluates the accuracy
         of the merged_model"""
 
         # compile the model
         (self.merged_model).compile(optimizer='adam', loss='binary_crossentropy', metrics=['acc'])
```
Comment (warmlogic) [inline Insight_Project_Framework/embeddings_DL.py:53]: Have you considered other metrics to compute like F1 score, which would help with an imbalanced dataset?

File: Insight_Project_Framework/data_processing.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    return np.transpose(X)
+
+def combine_rare(df,column, Limit = 200):
+    """Combine rare categorical data to Rare_"""
+    for r in df[column].unique():
+        if (np.sum(df[df[column] ==r]['Status'].value_counts()) < Limit):
+            df[column].replace(r,'Rare_'+column,inplace=True)
+
+def oversample_minority(df, ratio = 1.0):
```
Comment (warmlogic) [inline Insight_Project_Framework/data_processing.py:55]: How well does this method work for balancing the dataset? You could investigate others like SMOTE, etc

File: Insight_Project_Framework/data_processing.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    normalized_array -= amin
+    if center_at_zero:
+        normalized_array *= 2.0/(amax-amin)
+        normalized_array -= 1.0
+    else:
+        normalized_array *= 1.0/(amax-amin)
+    return normalized_array
+
+
+def dataframe_to_numpy(df,categorical_columns,ordinal_columns):
```
Comment (warmlogic) [inline Insight_Project_Framework/data_processing.py:33]: Just something to consider: how might you persist this encoder for use in the future if you had to evaluate new data?

File: Insight_Project_Framework/main.py [hunk 4]
```diff
@@ -161,26 +122,46 @@ def undersample_majority(df, ratio = 1.0):
...
+        Vis.data_exploration(df_clean)
+        Vis.country_vs_status(df_clean)
 
     # Split data set into training and test sets
     df_train_raw, df_test = train_test_split(df_clean, test_size=0.3 )
 
-    # Undersample the majority or oversample the minority
-    df_train = undersample_majority(df_train_raw,2)
-    #df_train = oversample_minority(df_train_raw)
+    # Undersample the majority (default) or ...
```
Comment (warmlogic) [inline Insight_Project_Framework/main.py:136]: How much data do you throw away when you undersample the majority?

Comment (warmlogic) [reply Insight_Project_Framework/main.py:136]: For the train test split above, you could stratify to get a good data sample (E.g. by country) if you want. Can set the random state if you want repeatable results.

**Review by warmlogic — APPROVED** [review]
Reviewing from my phone so apologies for brevity... This is looking good. I made a few comments. Have you considered tuning model (LR or RF) hyperparameters with random search CV or similar? I'm also curious about evaluation metrics other than accuracy. Seems like you're in a good spot though! Looking forward to seeing more about the embeddings.

File: Insight_Project_Framework/embeddings_DL.py (hunk 1 above)
Comment (michail-tzoufras) [reply Insight_Project_Framework/embeddings_DL.py:53]: I read up on this and decided to use the precision-recall curve as a metric. But the F1 score is also calculated in the new version.

File: Insight_Project_Framework/data_processing.py (hunk 3 above)
Comment (michail-tzoufras) [reply Insight_Project_Framework/data_processing.py:33]: Should I create a new "other" category and push new entries there?

File: Insight_Project_Framework/data_processing.py (hunk 2 above)
Comment (michail-tzoufras) [reply Insight_Project_Framework/data_processing.py:55]: I read up on this but haven't prioritized it yet.

File: Insight_Project_Framework/main.py (hunk 4 above)
Comment (michail-tzoufras) [reply Insight_Project_Framework/main.py:136]: I throw away > 90% of my training data. I will try other approaches too.

File: Insight_Project_Framework/data_processing.py (hunk 3 above)
Comment (warmlogic) [reply Insight_Project_Framework/data_processing.py:33]: I was thinking you could save the encoder as part of the model, or as a separate preprocessor object so all the data could be preprocessed in the same way. Not necessary for this project, but something to think about.

File: Insight_Project_Framework/data_processing.py (hunk 2 above)
Comment (warmlogic) [reply Insight_Project_Framework/data_processing.py:55]: here's an article to read about imbalanced classes, if you're curious https://www.svds.com/learning-imbalanced-classes/

File: Insight_Project_Framework/embeddings_DL.py (hunk 1 above)
Comment (warmlogic) [reply Insight_Project_Framework/embeddings_DL.py:53]: Won't you need to summarize the P-R curve as a single value? i.e., as the F1 score?

Comment (michail-tzoufras) [reply Insight_Project_Framework/embeddings_DL.py:53]: Oh yes, I do have the f1 score in the visualization.py in function 
"def plot_precision_recall(...)"
but it was commented out
But I decided in the last version (to be uploaded tomorrow) to report a single metric as you said. I had picked the True Default Rate (the recall for the "Defaulted" class). But perhaps I can report the f1 score instead. Is the f1 score more standard in the industry?

**Comment by warmlogic** [comment]
If I understand correctly, this is what you're asking the model to optimize. So currently it's optimizing accuracy, but you may want to use a custom metric like F1 if you care about making different kinds of mistakes more than others.

**Comment by michail-tzoufras** [comment]
Thanks, now I understand! I will look into changing that and will update you on the results.
